docs/src/tutorials/performance.jl

```diff
@@ -0,0 +1,158 @@
+
+# #### Minimise Runtime Exceptions
+
+# Many common operations can throw errors at runtime in Julia, they often do this by branching and calling a function in that branch both of which are slow on GPUs. Using `@inbounds` when indexing into arrays will eliminate exceptions due to bounds checking. You can also use `assume` from the package LLVM.jl to get rid of exceptions
+
+# ```julia
+# using LLVM, LLVM.Interop
+
+# function test(x, y)
+#     assume(x > 0)
+#     div(y, x)
+# end
+# ```
+
+# The `assume(x > 0)` tells the compiler that there cannot be a divide by 0 error.
+
+# For more information and examples check out [Kernel analysis and optimization](https://github.com/JuliaComputing/Training/blob/master/AdvancedGPU/2-2-kernel_analysis_optimization.ipynb).
+
+# #### 32-bit Integers
+
+# Use 32-bit integers where possible. A common source of register pressure is the use of 64-bit integers when only 32-bits are required. For example, the hardware's indices are 32-bit integers, but Julia's literals are Int64's which results in expressions like blockIdx().x-1 to be promoted to 64-bit integers. We can instead replace the `1` to a 32-bit integer with `Int32(1)` or more succintly `1i32` if you run `using CUDA: i32`
+
+# To see how much of a difference this actually makes lets use a kernel introduced in the introduction for inplace addition.
+
+using CUDA, BenchmarkTools
+
+function gpu_add3!(y, x)
+    index = (blockIdx().x - 1) * blockDim().x + threadIdx().x
+    stride = gridDim().x * blockDim().x
+    for i = index:stride:length(y)
+        @inbounds y[i] += x[i]
+    end
+    return
+end
+
+# Now let's benchmark it:
+
+x_d = CUDA.fill(1.0f0, 2^28)
+y_d = CUDA.fill(2.0f0, 2^28)
+
+function bench_gpu3!(y, x)
+    kernel = @cuda launch=false gpu_add3!(y, x)
+    config = launch_configuration(kernel.fun)
+    threads = min(length(y), config.threads)
+    blocks = cld(length(y), threads)
+
+    CUDA.@sync kernel(y, x; threads, blocks)
+end
+
+# ```julia
+# @btime bench_gpu3!($y_d, $x_d)
+# ```
+
+# ```
+#   76.202 ms (58 allocations: 3.73 KiB)
+# ```
+
+# Our kernel using 32-bit integers is below
+
+function gpu_add4!(y, x)
+    index = (blockIdx().x - Int32(1)) * blockDim().x + threadIdx().x
+    stride = gridDim().x * blockDim().x
+    for i = index:stride:length(y)
+        @inbounds y[i] += x[i]
+    end
+    return
+end
+
+function bench_gpu4!(y, x)
+    kernel = @cuda launch=false gpu_add4!(y, x)
+    config = launch_configuration(kernel.fun)
+    threads = min(length(y), config.threads)
+    blocks = cld(length(y), threads)
+
+    CUDA.@sync kernel(y, x; threads, blocks)
+end
+
+# ```julia
+# @btime bench_gpu4!($y_d, $x_d)
+# ```
+
+# ```
+#   76.149 ms (57 allocations: 3.70 KiB)
+# ```
+
+# As we can see from the benchmarks we only get a small performance benefit but there can be larger differences for more complex kernels.
```

Comment:
I don't think this measurement is really significant. Maybe instead show the amount of registers used by the kernel?